### PR TITLE
Extend SILCombiner code to handle existential self concrete type propagation using ProtocolConformanceAnalysis

### PR DESCRIPTION
**Pull request by rajbarik**
Extend SILCombiner code to handle existential self concrete type propagation using ProtocolConformanceAnalysis.

Prior PR:

### CODE REVIEW

**Comment by rajbarik** [comment]
@atrick  Could you review this one now?

**Comment by rajbarik** [comment]
@atrick @slavapestov Any possibility of getting this review done quicker? Thanks.

File: lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp [hunk 1]
```diff
@@ -603,6 +603,23 @@ SILCombiner::optimizeConcatenationOfStringLiterals(ApplyInst *AI) {
   return tryToConcatenateStrings(AI, Builder);
 }
 
+/// This routine replaces the old witness method inst with a new one.
+void SILCombiner::replaceWitnessMethodInst(
+    FullApplySite Apply, WitnessMethodInst *WMI, SILBuilderContext &BuilderCtx,
+    const CanType &ConcreteType, ProtocolConformanceRef &ConformanceRef) {
```
Comment (atrick) [inline lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp]: I think `ProtocolConformanceRef` should be passed by value.

File: lib/SILOptimizer/Utils/Existential.cpp [hunk 2]
```diff
@@ -249,14 +249,25 @@ SILInstruction *swift::findInitExistential(Operand &openedUse,
...
+/// If hasSoleConformingType is set, then the concrete type determination is
+/// deferred to a later stage where cast instructionas have been added.
+ConcreteExistentialInfo::ConcreteExistentialInfo(
+    Operand &openedUse, const bool hasSoleConformingType)
+    : hasSoleConformingType(hasSoleConformingType) {
+
+  //  Bail, if this existential has a sole conforming type. If there is a sole
+  //  confirming type, we defer the determination of concrete type until cast
+  //  instructions are added in SILCombiner.
+  if (hasSoleConformingType)
```
Comment (atrick) [inline lib/SILOptimizer/Utils/Existential.cpp]: It's strange that you would intentionally construct an invalid `ConcreteExistentialInfo`, presumably to "fill it in" later. I'll have to review the related patches to understand why you need to do this.

**Review by atrick — APPROVED** [review]
This look correct on its own, except for the couple comments above. I'm fine if this goes in and we can review the rest of the design elsewhere.

**Comment by rajbarik** [comment]
@atrick handled your concern, I think. Can you look at it again? Notably, I removed the InitExistential != null check in isValid(). This seems a redundant check given that we are checking everything else to be not null. Also, for using the same isValid() with SoleConformingType, we may not have an InitExistential. Please let me know more concerns soon. I will commit the other PR based on these changes.

**Comment by atrick** [comment]
LGTM

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Pull request merged**